vmm/src/device_manager.rs

```diff
@@ -980,20 +1049,38 @@ impl DeviceManager {
         pci: &mut PciBus,
         buses: &mut BusInfo,
         mem_slots: u32,
-    ) -> DeviceManagerResult<()> {
+    ) -> DeviceManagerResult<BTreeMap<u32, Arc<dyn ExternalDmaMapping>>> {
```

Comment:
So to me, the abstraction would be nicer if:

1) `add_vfio_devices` would take a `iommu_device: Option<vm_virtio::Iommu>` argument
2) Directly call into `iommu_device.add_mapping()` whenever we have an `iommu_device` and the VFIO device needs to be added to it.

It would be a little longer (Although I'm not expecting to pass dozens of VFIO devices to our virtio IOMMU), but cleaner imho. It would avoid having to call `iommu_device.set_external_mapping(vfio_iommu_device_mappings);`, which is VFIO specific, while setting the IOMMU device. In other words, it would make VFIO devices setting depend on the IOMMU device rather than the other way around.

Reply:
Yes I like this suggestion :)

Reply:
Done, but I still had to return a simple Vec of u32 to set the IORT table.